# perform hyperparameter tuning with Ridge and Lasso regression

In [1]:
import pandas as pd
import numpy as np

In [2]:
#import dataset
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()

In [3]:
# Load the California housing dataset
housing = fetch_california_housing()
data, target = housing.data, housing.target

In [4]:
housing

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [5]:
# Convert the data to a pandas DataFrame
housing_df = pd.DataFrame(data=housing.data, columns=housing.feature_names)

In [6]:
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [7]:
#independent feature and dependent feature 
x=housing_df
y=housing.target

In [8]:
# train test split
from sklearn.model_selection import train_test_split

In [9]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [10]:
X_train

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
14196  3.2596      33.0  5.017657   1.006421      2300.0  3.691814     32.71   
8267   3.8125      49.0  4.473545   1.041005      1314.0  1.738095     33.77   
17445  4.1563       4.0  5.645833   0.985119       915.0  2.723214     34.66   
14265  1.9425      36.0  4.002817   1.033803      1418.0  3.994366     32.69   
2271   3.5542      43.0  6.268421   1.134211       874.0  2.300000     36.78   
...       ...       ...       ...        ...         ...       ...       ...   
11284  6.3700      35.0  6.129032   0.926267       658.0  3.032258     33.78   
11964  3.0500      33.0  6.868597   1.269488      1753.0  3.904232     34.02   
5390   2.9344      36.0  3.986717   1.079696      1756.0  3.332068     34.03   
860    5.7192      15.0  6.395349   1.067979      1777.0  3.178891     37.58   
15795  2.5755      52.0  3.402576   1.058776      2619.0  2.108696     37.77   

       Longitude  
14196    -117.03  
8267     -118.16  
17445    -120.48  
14265    -117.11  
2271     -119.80  
...          ...  
11284    -117.96  
11964    -117.43  
5390     -118.38  
860      -121.96  
15795    -122.42  

[16512 rows x 8 columns]

In [11]:
y_train

array([1.03 , 3.821, 1.726, ..., 2.221, 2.835, 3.25 ])

In [12]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [13]:
##GridSearchCV
from sklearn.model_selection import train_test_split, GridSearchCV

In [14]:
alpha_values=[0.001, 0.01, 0.1, 1.0, 10.0, 100.0]

In [15]:
#Ridge Regression with GridSearchCV
from sklearn.linear_model import Ridge, Lasso
ridge_param_grid = [{'alpha': alpha_values}]
ridge_model = Ridge()
ridge_grid_search = GridSearchCV(ridge_model, ridge_param_grid, cv=5, scoring='neg_mean_squared_error')
ridge_grid_search.fit(X_train, y_train)
ridge_pred = ridge_grid_search.predict(X_test)

In [16]:
# Lasso Regression with GridSearchCV
lasso_param_grid = [{'alpha': alpha_values}]
lasso_model = Lasso()
lasso_grid_search = GridSearchCV(lasso_model, lasso_param_grid, cv=5, scoring='neg_mean_squared_error')
lasso_grid_search.fit(X_train, y_train)
lasso_pred = lasso_grid_search.predict(X_test)

In [17]:
# Evaluate Ridge Regression
from sklearn.metrics import mean_squared_error, r2_score
ridge_mse = mean_squared_error(y_test, ridge_pred)
ridge_r2 = r2_score(y_test, ridge_pred)

In [18]:
# Evaluate Lasso Regression
lasso_mse = mean_squared_error(y_test, lasso_pred)
lasso_r2 = r2_score(y_test, lasso_pred)

In [19]:
print("Ridge Regression:")
print(f"Best Alpha: {ridge_grid_search.best_params_['alpha']:.3f}")
print(f"Mean Squared Error: {ridge_mse:.2f}")
print(f"R-squared (R2) Score: {ridge_r2:.2f}")
print("\nLasso Regression:")
print(f"Best Alpha: {lasso_grid_search.best_params_['alpha']:.3f}")
print(f"Mean Squared Error: {lasso_mse:.2f}")
print(f"R-squared (R2) Score: {lasso_r2:.2f}")

Ridge Regression:
Best Alpha: 0.100
Mean Squared Error: 0.56
R-squared (R2) Score: 0.58

Lasso Regression:
Best Alpha: 0.001
Mean Squared Error: 0.55
R-squared (R2) Score: 0.58


The results indicate the performance of Ridge and Lasso regression models with the best hyperparameters found using GridSearchCV. Here's what the results mean:

**Ridge Regression:**
- Best Alpha: 0.100
  - This is the best hyperparameter (alpha) found by GridSearchCV, which balances the trade-off between model complexity and fit to the data. In this case, the best alpha is 0.100.
- Mean Squared Error: 0.56
  - The mean squared error (MSE) measures the average squared difference between predicted and actual values. In this Ridge model, the MSE is approximately 0.56.
- R-squared (R2) Score: 0.58
  - The R-squared (R2) score indicates the proportion of variance in the dependent variable (house prices) explained by the independent variables. An R2 score of 0.58 suggests that approximately 58% of the variance in house prices is explained by the features in the Ridge model.

**Lasso Regression:**
- Best Alpha: 0.001
  - In the Lasso model, the best alpha found by GridSearchCV is 0.001, indicating the optimal level of regularization.
- Mean Squared Error: 0.55
  - The mean squared error (MSE) for the Lasso model is approximately 0.55, which is slightly lower than the Ridge model's MSE.
- R-squared (R2) Score: 0.58
  - The R-squared (R2) score for the Lasso model is also approximately 0.58, indicating a similar level of variance explained by the model as the Ridge model.

These results show that both Ridge and Lasso regression models perform similarly in terms of R2 score and MSE on this dataset. Lasso's best alpha is smaller, suggesting that it utilizes less regularization, and its MSE is slightly lower than Ridge's, indicating a marginally better fit to the data. However, the differences are relatively small, and the choice between Ridge and Lasso may depend on factors like interpretability, feature selection, and domain-specific considerations.